### ISSUE TITLE
Kernel dispatch & computation call flow (B200 / SM100)

### ISSUE THREAD

**Issue by yiliu30** [thu-ml/SpargeAttn#120]
# SpargeAttn Kernel Dispatch & Computation

## Architecture Dispatch

The Python entry points (`spas_sage_attn/core.py`) select block sizes and kernel variants based on `torch.cuda.get_device_capability()`:

| Arch | GPU | Block Sizes (Q,K) | V dtype | Kernel |
|------|-----|--------------------|---------|--------|
| sm80/86/87 | A100, A40, A6000 | 128, 64 | FP16 | `qk_int_sv_f16_cuda_sm80.cuh` |
| sm89 | L40S, RTX 6000 Ada | 128, 64 | FP8 | `qk_int_sv_f8_cuda_sm89.cuh` |
| sm90 | H100, H200 | 64, 128 | FP8 | `qk_int_sv_f8_cuda_sm90.cuh` (WGMMA) |
| sm100 | B100, B200 | 128, 64 | FP8 | `qk_int_sv_f8_cuda_sm89.cuh` |

B200 (SM100) uses the SM89 kernel path because no SM100-specific kernel exists. The SM89 kernel compiles cleanly for SM100 with nvcc 13.2.

## Call Flow (B200 / SM100)

```
vllm-qdq-plugin: impl.py
  SpargeAttnImpl._forward_sparge()
    │
    │ mode == "topk"?
    ├── spas_sage2_attn_meansim_topk_cuda()        # topk-based block selection
    └── spas_sage2_attn_meansim_cuda()             # cdfthreshd-based block selection
         │
         ▼
    spas_sage_attn/core.py
      │
      ├── 1. get_block_map_meansim_fuse_quant()    # [spas_sage_attn/utils.py]
      │       ├── get_pool_sim_triton_simmean_fuse_quant(Q)  → pool_Q, sim_Q, Q_int8, Q_scale
      │       ├── get_pool_sim_triton_simmean_fuse_quant(K)  → pool_K, sim_K, K_int8, K_scale
      │       ├── pool_Q @ pool_K^T / sqrt(d)                → block-level proxy scores
      │       ├── softmax → sort → cdf → searchsorted/cut    → num_to_select per Q-row
      │       ├── fill_block_map_triton()                     → boolean mask [nQ,nK]
      │       └── block_map_lut_triton()                      → lut, valid_block_num
      │
      ├── 2. fused.transpose_pad_permute_cuda(V)   # V: [B,H,N,D] → [B,H,D,padded_N]
      ├── 3. fused.scale_fuse_quant_cuda(V)        # V_fp8: quantize to FP8 E4M3
      │
      └── 4. qattn.qk_int8_sv_f8_accum_f32_..._with_pv_threshold()
              │                                    # [csrc/qattn/pybind.cpp]
              ▼
          csrc/qattn/qk_int_sv_f8_cuda_sm89.cu
            │
            │  CHECK_*  →  validate tensor shapes & strides
            │  DISPATCH_HEAD_DIM  →  instantiate for head_dim=64 or 128
            │  DISPATCH_CTA_SIZE  →  CTA_Q=128, CTA_K=64
            │  DISPATCH_IS_CAUSAL →  is_causal=false
            │
            └── SpargeAttentionSM89Dispatched<CTA_Q, CTA_K, ...>()
                    │                            # [qk_int_sv_f8_cuda_sm89.cuh:757]
                    │
                    │  template params:
                    │    CTA_Q=128, CTA_K=64, WARP_Q=64, WARP_K=64
                    │    head_dim=64 or 128, qk_quant_gran=kPerWarp
                    │    DTypePVAccum=float, use_inst_buffer=true
                    │    PVThresholdMode=kPerBlock, DTypeOut=half
                    │    DenominatorAccumUnit=kCudaCore
                    │    mask_mode=kNone (or kCausal), fuse_v_scale=true
                    │
                    └── qk_int_sv_f8_block_sparse_attn_kernel<<<grid, block, smem>>>
                            │                        # [qk_int_sv_f8_cuda_sm89.cuh:44]
                            │
                            │  Per KV-block iteration (lut-driven):
                            │
                            ├── compute_int_qk()    # attn_utils.cuh:248
                            │     ├── ldmatrix_m8n8x4  (load Q tile from smem)
                            │     ├── ldmatrix_m8n8x4  (load K tile from smem)
                            │     └── mma::mma_sync_m16n16k32_row_col_s8s8s32
                            │           └── mma.sync.aligned.m16n8k32.row.col.s32.s8.s8.s32  (×2)
                            │                                  # mma.cuh:372
                            │
                            ├── update_mo()          # attn_utils.cuh:419
                            │     └── FP32 online softmax: max reduce, exp2 rescale
                            │
                            ├── [pvthreshd gate]    # skip P@V if max_diff + pv_threshold ≤ 0
                            │
                            ├── RS_32_to_8()         # convert softmax output → FP8
                            │
                            └── compute_fp8_sv_inst_buf()   # attn_utils.cuh:853
                                  ├── ldmatrix_m8n8x4  (load V tile from smem)
                                  └── mma::mma_sync_m16n16k32_row_col_f8f8f32
                                        └── mma.sync.aligned.m16n8k32.row.col.f32.e4m3.e4m3.f32  (×2)
                                                       # mma.cuh:630

                            ── normalize_d() → store O to global memory
```

## Key Source Files

| File | Role |
|------|------|
| `spas_sage_attn/core.py` | Python entry: block map, quantization, kernel dispatch |
| `spas_sage_attn/utils.py` | Block map construction (pool, sim, proxy attention, LUT) |
| `csrc/qattn/qk_int_sv_f8_cuda_sm89.cu` | SM89/SM100 host dispatch + validation |
| `csrc/qattn/qk_int_sv_f8_cuda_sm89.cuh` | SM89 kernel: INT8 QK + FP8 SV + pvthreshd + online softmax |
| `csrc/qattn/attn_utils.cuh` | `compute_int_qk`, `compute_fp8_sv`, `update_mo`, `update_mdo` |
| `csrc/mma.cuh` | PTX inline asm wrappers for `mma.sync` instructions |
| `csrc/qattn/pybind.cpp` | PyTorch ↔ CUDA bindings |

Each attention head computes `softmax(Q @ K^T / sqrt(d)) @ V`. SpargeAttn splits this into two tensor-core matmuls:

### 1. Q @ K^T (first matmul)

| | dtype | MMA tile | accumulator |
|---|---|---|---|
| Q | INT8, per-block quantized | | |
| K | INT8, per-block quantized | 16×16×32 | INT32 |
| S | — | | → FP32 for softmax |

```
Q_int8, Q_scale = per_block_quantize(Q)       # scale: max(|block|) / 127
K_int8, K_scale = per_block_quantize(K - km)   # km = K.mean(-2) if smooth_k

S_int32 = mma_int8(Q_int8, K_int8)             # INT8 tensor core
S_fp32  = int2float(S_int32) * Q_scale * K_scale * log2(e) / sqrt(d)
```

The `log2(e)` factor converts softmax from natural to base-2 exponent for numerical efficiency.

### 2. P @ V (second matmul)

| | dtype | MMA tile | accumulator |
|---|---|---|---|
| P | FP8 (E4M3) | | |
| V | FP8 (E4M3), pre-quantized | 16×16×32 | FP32 |
| O | — | | FP16 or BF16 |

```
# Softmax in FP32 (online algorithm, per KV block):
#   m_new = max(m_old, row_max(S_fp32))
#   P_fp32 = exp2(S_fp32 - m_new)
#   d = d * exp2(m_old - m_new) + row_sum(P_fp32)
#   O = O * exp2(m_old - m_new) + P_fp32 @ V_fp8

P_fp8 = fp32_to_fp8(P_fp32)                    # RS_32_to_8 — cast softmax output to FP8
O_fp32 += mma_fp8(P_fp8, V_fp8) * V_scale      # FP8 tensor core, fused V dequant
O_final = O_fp32 / d                            # normalize by denominator
```

V is quantized once per forward pass (before the kernel):
```
V_hnd → transpose + pad + permute → V_permuted
V_permuted → scale_fuse_quant → V_fp8, V_scale
```

The V dequant scale is fused into the S·V matmul, avoiding a separate dequant step. On SM80 (no FP8), V stays in FP16 and the second matmul uses FP16 tensor cores instead.

## Quantization Granularity

Q and K are quantized per-block (one scale per `BLKQ=128` or `BLKK=64` tokens). The kernel supports three granularity levels:

| Mode | Scale per | Typical use |
|------|-----------|-------------|
| kPerBlock | 1 scale per CTA tile | Default for all archs |
| kPerWarp | 1 scale per warp tile | Higher precision |
| kPerThread | 1 scale per 4-thread group | Maximum precision |

Current dispatch uses `qk_quant_gran=1` (kPerWarp).

## pvthreshd Skip Logic

After computing QK for a KV block, the kernel can skip the softmax + S·V step:

```
local_max_diff = max(row_max(S) - running_max)   // within warp/block
if local_max_diff + pv_threshold <= 0:
    skip P @ V for this block
```

This is a second compute-skip layer on top of the coarse block-map sparsity.

**Comment by yiliu30** [comment]
Created by accident. Sorry for the trouble — closing this now.
